### PR TITLE
Story/vspc 69

### PR DESCRIPTION
**Pull request by Sagar-k93**

### CODE REVIEW

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ImageService.java [hunk 1]
```diff
@@ -100,7 +100,32 @@ public ImageData getImageDimensions(IVSImage image, int width, int height) {
...
+     * @param pageNo. if pageNo<1, 1st page is returned, if pageNo>total pages,last
+     *                page is returned
+     * @return list of images in the requested pageNo and requested order.
+     */
+    @Override
+    public List<IVSImage> getImages(int pageNo, String sortedBy, String order) {
+        Sort sortingParameters = Sort.by(SortByField.CREATION_DATE.getValue()).descending();
+        pageNo = validatePageNumber(pageNo);
+        if (sortedBy != null) {
+            if (order.contentEquals("desc")) {
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ImageService.java]: if strings are hard-coded they should be constants

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ImageService.java [hunk 2]
```diff
@@ -100,7 +100,32 @@ public ImageData getImageDimensions(IVSImage image, int width, int height) {
...
+                sortingParameters = Sort.by(sortedBy).descending();
+            } else if(order.contentEquals("asc")) {
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ImageService.java]: in any other case it should be asc

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ImageService.java [hunk 3]
```diff
@@ -100,7 +100,32 @@ public ImageData getImageDimensions(IVSImage image, int width, int height) {
...
+     *                page is returned
+     * @return list of images in the requested pageNo and requested order.
+     */
+    @Override
+    public List<IVSImage> getImages(int pageNo, String sortedBy, String order) {
+        Sort sortingParameters = Sort.by(SortByField.CREATION_DATE.getValue()).descending();
+        pageNo = validatePageNumber(pageNo);
+        if (sortedBy != null) {
+            if (order.contentEquals("desc")) {
+                sortingParameters = Sort.by(sortedBy).descending();
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ImageService.java]: you probably want to validagte `sortBy` to make sure it's a valid field. For now, maybe just have a hard-coded list of valid sortable fields.

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/ListImagesController.java [hunk 4]
```diff
@@ -19,7 +20,7 @@
     private IImageService imageService;
 
     @RequestMapping("/staff/images/list/{page}")
-    public String listSpaces(@PathVariable String page, Model model) {
+    public String listSpaces(@PathVariable String page, @RequestParam(value = "sort", required = false) String sortedBy, @RequestParam(value = "order", required = false) String order, Model model) {
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/ListImagesController.java]: since you're working on this, let's change the method name. that doesn't sound right.

File: vspace/src/main/webapp/WEB-INF/views/staff/images/imagelist.jsp [hunk 5]
```diff
@@ -1,12 +1,13 @@
 <%@ page pageEncoding="UTF-8"%>
 <%@ taglib uri="http://java.sun.com/jsp/jstl/core" prefix="c"%>
 <%@ taglib prefix="sec"
-    uri="http://www.springframework.org/security/tags"%>
+	uri="http://www.springframework.org/security/tags"%>
 <%@ taglib prefix="spring" uri="http://www.springframework.org/tags"%>
 
 <script src="<c:url value="/resources/bootpag/js/bootpag.min.js" />"></script>
-<script>
+<script th:inline="javascript">
```
Comment (jdamerow) [inline vspace/src/main/webapp/WEB-INF/views/staff/images/imagelist.jsp]: why did you add that? isn't that a thymeleaf tag?

File: vspace/src/main/webapp/WEB-INF/views/staff/images/imagelist.jsp [hunk 6]
```diff
@@ -51,74 +74,64 @@ $( document ).ready(function() {
...
+									<option>Assign a tag</option>
+									<c:forEach items="${imageCategories}" var="catTag">
+										<option value="${catTag}"><spring:eval
+												expression="@configFile.getProperty('image_category_' + catTag)" /></option>
+									</c:forEach>
+								</select>
+							</form></td>
+					</tr>
+				</c:forEach>
+			</tbody>
```
Comment (jdamerow) [inline vspace/src/main/webapp/WEB-INF/views/staff/images/imagelist.jsp]: that shouldn't be indented that much

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ImageService.java [hunk 7]
```diff
@@ -82,92 +84,120 @@ public ImageData getImageDimensions(IVSImage image, int width, int height) {
...
+     * @param pageNo. if pageNo<1, 1st page is returned, if pageNo>total pages,last
+     *                page is returned
+     * @return list of images in the requested pageNo and requested order.
+     */
+    @Override
+    public List<IVSImage> getImages(int pageNo, String sortedBy, String order) {
+        Sort sortingParameters = Sort.by(SortByField.CREATION_DATE.getValue()).descending();
+        pageNo = validatePageNumber(pageNo);
+        if (sortedBy != null && !sortedBy.equals("") && SortByField.getAllValues().contains(sortedBy)) {
+            if (order.equalsIgnoreCase(Sort.Direction.DESC.toString())) {
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ImageService.java]: if only order is given but not sortedBy, we probably still want to use follow the order. Right now it is only used when sortBy is provided.

**Comment by jdamerow** [comment]
Jenkins, test this please.

**Comment by jdamerow** [comment]
on first page load it, the sorting indicator should be shown (i.e. asc by date)

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ImageService.java (hunk 3 above)
Comment (Sagar-k93) [reply vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ImageService.java]: The method SortByField.getAllValues().contains(sortedBy) at line 117 does validate if the sortedBy value is one of the predefined values for sorting.

Comment (jdamerow) [reply vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ImageService.java]: yes, I believe my comment was for the previous version

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ImageService.java [hunk 8]
```diff
@@ -75,99 +77,130 @@ public ImageData getImageDimensions(IVSImage image, int width, int height) {
...
+    /**
+     * Method to return the requested images
+     * 
+     * @param pageNo. if pageNo<1, 1st page is returned, if pageNo>total pages,last
+     *                page is returned
+     * @return list of images in the requested pageNo and requested order.
+     */
+    @Override
+    public List<IVSImage> getImages(int pageNo, String sortedBy, String order) {
+        System.out.println("inside getImages== " + sortedBy + " -- " + order);
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ImageService.java]: remove

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ImageService.java [hunk 9]
```diff
@@ -75,99 +77,130 @@ public ImageData getImageDimensions(IVSImage image, int width, int height) {
...
+        Sort sortingParameters = Sort.by(SortByField.CREATION_DATE.getValue()).descending();
+        pageNo = validatePageNumber(pageNo);
+        if(sortedBy!=null && !sortedBy.equals("") && SortByField.getAllValues().contains(sortedBy)) {
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ImageService.java]: you can remove `!sortedBy.equals("")`; that should be checked through the `contains` call on allvalues.

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ImageService.java [hunk 10]
```diff
@@ -75,99 +77,130 @@ public ImageData getImageDimensions(IVSImage image, int width, int height) {
...
+            if(order.equalsIgnoreCase(Sort.Direction.DESC.toString())) {
+                sortingParameters = Sort.by(sortedBy).descending();
+            } else {
+                sortingParameters = Sort.by(sortedBy);
+            }
+        } else if(order!=null && !order.equals("") && order.equalsIgnoreCase(Sort.Direction.ASC.toString())) {
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ImageService.java]: `!order.equals("")` is check in `order.equalsIgnoreCase(Sort.Direction.ASC.toString())`

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ImageService.java [hunk 11]
```diff
@@ -75,99 +77,130 @@ public ImageData getImageDimensions(IVSImage image, int width, int height) {
...
+     *                page is returned
+     * @return list of images in the requested pageNo and requested order.
+     */
+    @Override
+    public List<IVSImage> getImages(int pageNo, String sortedBy, String order) {
+        System.out.println("inside getImages== " + sortedBy + " -- " + order);
+        Sort sortingParameters = Sort.by(SortByField.CREATION_DATE.getValue()).descending();
+        pageNo = validatePageNumber(pageNo);
+        if(sortedBy!=null && !sortedBy.equals("") && SortByField.getAllValues().contains(sortedBy)) {
+            if(order.equalsIgnoreCase(Sort.Direction.DESC.toString())) {
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ImageService.java]: order is checked in `else if`; just make sure it's not an `else if` but and `if`

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/ListImagesController.java [hunk 12]
```diff
@@ -1,38 +1,44 @@
...
+		model.addAttribute("totalImageCount", imageService.getTotalImageCount());
+		model.addAttribute("images", imageService.getImages(pageNo, sortedBy, order));
+		model.addAttribute("imageCategories", ImageCategory.values());
+		model.addAttribute("sortProperty",
+				(sortedBy == null || sortedBy.equals("")) ? SortByField.CREATION_DATE.getValue() : sortedBy);
+		model.addAttribute("order",
+				(order == null || order.equals("")) ? Sort.Direction.DESC.toString().toLowerCase() : order);
+
+		return "staff/images/list";
+	}
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/ListImagesController.java]: indentation!!!!!

File: vspace/src/main/webapp/WEB-INF/views/staff/images/imagelist.jsp (hunk 6 above)
Comment (jdamerow) [reply vspace/src/main/webapp/WEB-INF/views/staff/images/imagelist.jsp]: ?

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ImageService.java (hunk 11 above)
Comment (jdamerow) [reply vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ImageService.java]: why do you need this? just extend the if in 124 with an else.

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ImageService.java [hunk 13]
```diff
@@ -75,99 +77,129 @@ public ImageData getImageDimensions(IVSImage image, int width, int height) {
...
      * 
-     * @param pageNo. if pageNo<1, 1st page is returned, if pageNo>total pages,last page is returned
+     * @param pageNo. if pageNo<1, 1st page is returned, if pageNo>total pages,last
+     *                page is returned
      * @return list of images in the requested pageNo
-     */ 
+     */
     @Override
-    public List<IVSImage > getImages(int pageNo) {
+    public List<IVSImage> getImages(int pageNo) {
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ImageService.java:96]: this method does the same thing as below `getImages` method, doesn't it? except using default values? it should call the other method to avoid duplicate code.

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ImageService.java [hunk 14]
```diff
@@ -75,99 +78,123 @@ public ImageData getImageDimensions(IVSImage image, int width, int height) {
...
+
+    /**
+     * Method to return the requested images
+     * 
+     * @param pageNo. if pageNo<1, 1st page is returned, if pageNo>total pages,last
+     *                page is returned
+     * @return list of images in the requested pageNo and requested order.
+     */
+    @Override
+    public List<IVSImage> getImages(int pageNo, String sortedBy, String order) {
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ImageService.java:109]: needs a test case

**Comment by jdamerow** [comment]
after that should be good to go

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ImageServiceTest.java [hunk 15]
```diff
@@ -150,6 +150,38 @@ public void test_getImages_pageGreaterThanTotalPages() {
...
     
+    @Test
+    public void test_getImages_sorted_success() { 
+        Pageable sortByRequestedField = PageRequest.of(0, 10, Sort.by(SortByField.CREATION_DATE.getValue()).descending());
+        when(imageRepo.count()).thenReturn(1L);
+        when(imageRepo.findAll(sortByRequestedField)).thenReturn(new PageImpl<VSImage>(images));
+        List<IVSImage> requestedImages = serviceToTest.getImages(1, SortByField.CREATION_DATE.getValue(), Sort.Direction.DESC.toString());
+        assertEquals(1, requestedImages.size());
+        assertEquals(IMG_ID, requestedImages.get(0).getId());
+        verify(imageRepo).findAll(sortByRequestedField);
```
Comment (jdamerow) [inline vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ImageServiceTest.java:161]: since you have a return value to test if the method does what it's supposed, I wouldn't verify this. There could be other ways of getting the correct results that do not call `findAll` and by putting this here, your test case will break if we change the implementation.

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ImageServiceTest.java [hunk 16]
```diff
@@ -150,6 +150,38 @@ public void test_getImages_pageGreaterThanTotalPages() {
...
+    }
+    
+    @Test
+    public void test_getImages_sorted_negativePage() { 
+        Pageable sortByRequestedField = PageRequest.of(0, 10, Sort.by(SortByField.CREATED_BY.getValue()).descending());
+        when(imageRepo.count()).thenReturn(1L);
+        when(imageRepo.findAll(sortByRequestedField)).thenReturn(new PageImpl<VSImage>(images));
+        List<IVSImage> requestedImages = serviceToTest.getImages(-2, SortByField.CREATED_BY.getValue(), Sort.Direction.DESC.toString());
+        assertEquals(IMG_ID, requestedImages.get(0).getId());
+        verify(imageRepo).findAll(sortByRequestedField);
```
Comment (jdamerow) [inline vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ImageServiceTest.java:171]: same above

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ImageServiceTest.java [hunk 17]
```diff
@@ -150,6 +150,38 @@ public void test_getImages_pageGreaterThanTotalPages() {
...
+    
+    @Test
+    public void test_getImages_sorted_pageGreaterThanTotalPages() { 
+        ReflectionTestUtils.setField(serviceToTest, "pageSize", 1);
+        Pageable sortByRequestedField = PageRequest.of(4, 1, Sort.by(SortByField.CREATION_DATE.getValue()).descending());
+        when(imageRepo.count()).thenReturn(5L);
+        when(imageRepo.findAll(sortByRequestedField)).thenReturn(new PageImpl<VSImage>(images));
+        List<IVSImage> requestedImages = serviceToTest.getImages(7, SortByField.CREATION_DATE.getValue(), Sort.Direction.DESC.toString());
+        assertEquals(IMG_ID, requestedImages.get(0).getId());
+        verify(imageRepo).findAll(sortByRequestedField);
```
Comment (jdamerow) [inline vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ImageServiceTest.java:182]: same as above

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ImageServiceTest.java [hunk 18]
```diff
@@ -150,6 +150,38 @@ public void test_getImages_pageGreaterThanTotalPages() {
...
+    }
+
```
Comment (jdamerow) [inline vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ImageServiceTest.java:184]: a test case for if there are no results is missing

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ImageService.java [hunk 19]
```diff
@@ -75,99 +76,125 @@ public ImageData getImageDimensions(IVSImage image, int width, int height) {
...
+            sortingParameters = sortingParameters.ascending();
+        }
+        else {
+            sortingParameters = sortingParameters.descending();
+        }
+        Pageable sortByRequestedField = PageRequest.of(pageNo - 1, pageSize, sortingParameters);
         Page<VSImage> images = imageRepo.findAll(sortByRequestedField);
         List<IVSImage> results = new ArrayList<>();
-        images.getContent().forEach(i -> results.add(i));
+        if(!java.util.Objects.isNull(images)) {
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ImageService.java:122]: why are you using this instead of `== null`?

**Comment by jdamerow** [comment]
Jenkins, test this please.

**Comment by jdamerow** [comment]
after that it should be fine

**Comment by jdamerow** [comment]
Jenkins, test this please.
